orangecontrib/text/widgets/owpreprocess.py

```diff
@@ -1,70 +1,706 @@
+        frame_layout = QVBoxLayout()
+        frame_layout.setContentsMargins(0, 0, 0, 0)
+        frame_layout.setSpacing(0)
+        frame.setLayout(frame_layout)
+        # Load the previous states.
+        persistent_data = [
+            self.persistent_data_tokenizer,
+            self.persistent_data_casing,
+            self.persistent_data_stemmer,
+            self.persistent_data_filter,
+        ]
+        for ModuleClass, ModuleData in zip(PREPROCESSOR_MODULES,
+                                           persistent_data):
+            pp_module_widget = ModuleClass(ModuleData)  # Create pp instance.
+            self.preprocessors.append(pp_module_widget)
+            pp_module_widget.change_signal.connect(self.settings_invalidated)
+            pp_module_widget.error_signal.connect(self.display_message)
 
-        self.trans_combo = QtGui.QComboBox(settings_box)
-        transformation_box.layout().addWidget(self.trans_combo, QtCore.Qt.AlignRight)
-        self.fill_transformation_options()  # Add available options.
-        self.trans_combo.activated[int].connect(self.select_transformation)
+            frame_layout.addWidget(pp_module_widget)
+        self.store_pipeline()  # Store the pipeline after loading it.
 
-        gui.button(self.controlArea, self, "&Apply", callback=self.apply, default=True)
+        frame_layout.addStretch()
+        self.mainArea.layout().addWidget(frame)
 
-        self.apply()
+    def set_data(self, data=None):
+        self.corpus = data
+        self.update_info()
+        self.commit()
 
-    def select_transformation(self, n):
-        if n > 0:
-            self.transformation = self.transformation_opt[n]
+    def update_info(self):
+        if self.corpus is not None:
+            info = 'Document count: {}'.format(len(self.corpus))
         else:
-            self.transformation = None
-
-    def fill_transformation_options(self):
-        self.trans_combo.clear()
-        for trans in self.transformation_opt:
-            if trans == "(none)":
-                self.trans_combo.addItem("(none)")
-            else:
-                self.trans_combo.addItem(trans)
-
-    def apply(self):
-        # TODO change this to custom stopwords
-        if self.remove_stpwrds:
-            sw = 'english'
-        pp = Preprocessor(incl_punct=self.include_punctuation, trans=self.transformation,
-                          lowercase=self.lowercase, stop_words=sw)
-        self.send(Output.PREPROCESSOR, pp)
\ No newline at end of file
+            info = 'No input corpus detected.'
+        self.info_label.setText(info)
+
+    def commit(self):
+        self.store_pipeline()  # Store the new pipeline.
+        if self.corpus is not None:
+            pp = self.assemble_preprocessor()
+            if pp is not None:
+                self.apply(pp)
+
+    def apply(self, preprocessor):
+        output = preprocessor(self.corpus)
+        self.send(Output.PP_CORPUS, output)
```

Comment:
The output shoud be a `Corpus`. I think, we should store `tokens` into `Corpus.tokens` and then send out the `Corpus`, not `list`.